File: Python3/404. Sum of Left Leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
#         sum = [0] # 44ms
#         def recursion(node, sum, isLeft):
#             if node:
#                 if isLeft and not node.left and not node.right: sum[0] += node.val
#                 recursion(node.left, sum, True)
#                 recursion(node.right, sum, False)
                
#         recursion(root, sum, False)
#         return sum[0]

        if not root: # 48ms
            return 0
        else:
            if root.left and not root.left.left and not root.left.right:
                return root.left.val + self.sumOfLeftLeaves(root.right)
            else:
                return self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)

File: Golang/404. Sum of Left Leaves.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func sumOfLeftLeaves(root *TreeNode) int { // 4ms
//     sum := 0
//     recursion(root, &sum, false)
//     return sum
// }

// func recursion(node *TreeNode, sum *int, isLeft bool) {
//     if node != nil {
//         if node.Left == nil && node.Right == nil && isLeft {
//             *sum += node.Val
//         }
//         recursion(node.Left, sum, true)
//         recursion(node.Right, sum, false)
//     }
// }

func sumOfLeftLeaves(root *TreeNode) int {
    if root == nil {
        return 0
    } else if root.Left != nil && root.Left.Left == nil && root.Left.Right == nil {
        return root.Left.Val + sumOfLeftLeaves(root.Right)
    } else {
        return sumOfLeftLeaves(root.Left) + sumOfLeftLeaves(root.Right)
    }
}
